pyccel/parser/semantic.py

```diff
@@ -409,27 +408,23 @@ def _infere_type(self, expr, **settings):
             return d_var
 
         elif isinstance(expr, Variable):
-
             d_var['datatype'      ] = expr.dtype
-            d_var['allocatable'   ] = expr.allocatable if expr.rank>0 else False
+            d_var['memory_handling'] = 'heap' if expr.rank > 0 else expr.memory_handling
```

Comment:
The `rank` is the number of dimensions so a 0D object is a scalar and anything else is an array.

Before we had:
```
if array:
      d_var['allocatable'   ] = expr.allocatable # (stack or heap to match rhs)
else:
     d_var['memory_handling'   ] = 'stack'
```

Now we have:
```
if array:
     d_var['memory_handling'   ] = 'heap'
else:
    d_var['memory_handling'   ] = expr.memory_handling
```

Actually given that scalars now have a `memory_handling` property maybe this can just be:
```
d_var['memory_handling'   ] = expr.memory_handling
```